File: Part1_강의자료/Part1_연습문제_풀이/1713_후보추천하기/1713.py
class student:
    def __init__(self, number, scr, posted_at):
        self.number = number
        self.scr = scr
        self.posted_at = posted_at

N = int(input())
R = int(input())
voted_number_list = list(map(int,input().split()))
post = []

for i in range(R):
    already_posted = False
    
    for j in range(len(post)):
        if post[j].number == voted_number_list[i]:
            post[j].scr += 1
            already_posted = True
            break
    
    if not already_posted:
        if len(post) < N:
            post.insert(0, student(voted_number_list[i], 1, i))
        else:
            post.sort(key=lambda s: (s.scr, s.posted_at))
            post[0] = student(voted_number_list[i], 1, i)

post.sort(key=lambda s: s.number)
print(' '.join(str(s.number) for s in post))

File: Part1_강의자료/Part1_연습문제_풀이/1713_후보추천하기/1713.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

struct student {
    int number;
    int scr;
    int posted_at;
};

int main()
{
    int N, R;
    vector<student> post;

    scanf("%d", &N);
    scanf("%d", &R);
    for (int i = 0; i < R; i++)
    {
        int voted_number;
        scanf("%d", &voted_number);

        bool already_posted = false;
        for (int j = 0; j < post.size(); j++)
        {
            if (post[j].number == voted_number)
            {
                post[j].scr++;
                already_posted = true;
                break;
            }
        }

        if (!already_posted)
        {
            if (post.size() < N)
                post.push_back({voted_number, 1, i});
            else
            {
                sort(post.begin(), post.end(), [](student o1, student o2)->bool{ return o1.scr == o2.scr ? o1.posted_at < o2.posted_at : o1.scr < o2.scr; });
                post[0] = {voted_number, 1, i};
            }
        }
    }

    sort(post.begin(), post.end(), [](student o1, student o2)->bool{ return o1.number < o2.number; });
    for (int i = 0; i < post.size(); i++)
        printf("%d ", post[i]);
    printf("\n");
    return 0;
}
